Expose a public API for building RQNames

Pilchie opened the pull request:
* Rename RQNameService to RQNameBuilder and make public.
* Rename TryBuildForPublicApi to Build and have it return null.
* Remove TryBuild, since we *only* build RQNames for public Api's now.
* Clean up implementation to not pass flag around for buildForPublicApi
* Remove RQDynamicType
* Create a singleton for System.Object.
* Add a comment explaining why we erase dynamic.

Fixes #55.

Pilchie commented:
@mattwar - are you happy with the API shape?
@dpoeschl - okay with the simplification around `buildForPublicAPI`

mattwar commented:
The type should be called RQName, and the method should be called Create or From.
It should not appear in a namespace called Implementation.
Is there not also an API that converts back to a symbol?

Pilchie commented:
Thanks!  Renamed `Build` to `From` and moved to the root `Microsoft.VisualStudio.LanguageServices` namespaces.

There is no API that converts back, because we don't ever take RQNames as input - we just give them out when we are doing rename/go to definition.

Pilchie commented:
Not saying we couldn't/shouldn't add a method to go back to `ISymbol`, just that we don't have one right now.

mattwar commented:
:+1:

dpoeschl commented on src/VisualStudio/Core/Def/Implementation/RQName/Nodes/RQType.cs:
```diff
@@ -1,11 +1,16 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using Microsoft.VisualStudio.LanguageServices.Implementation.RQName.SimpleTree;
+using System;
```
Wrong sorting?

Pilchie replied:
Yep, fixed

dpoeschl commented on src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs:
```diff
@@ -1,25 +1,23 @@
 {
-    internal static class RQNameService
+    /// <summary>
+    /// Helpers related to <see cref="RQName"/>s.
```
Should we explain what RQNames are here (and cref to IVsRefactorNotify)? It'd be nice if the Metadata view of IVsRefactorNotify included the remarks from the [msdn docs](https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.interop.ivsrefactornotify.aspx) with the grammar and examples.

Pilchie replied:
Added the crefs.

dpoeschl commented on src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs:
```diff
@@ -1,25 +1,23 @@
+    /// <summary>
+    /// Helpers related to <see cref="RQName"/>s.
+    /// </summary>
+    public static class RQName
```
Update the file name to match

Pilchie replied:
Done.

dpoeschl commented on src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs:
```diff
@@ -1,25 +1,23 @@
-            return node != null;
-        }
-
-        public static bool TryBuildForPublicAPIs(ISymbol symbol, out string rqname)
```
Very happy to see this go. :smile:

dpoeschl commented:
:+1:
